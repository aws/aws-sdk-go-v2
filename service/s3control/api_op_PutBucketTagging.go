// Code generated by smithy-go-codegen DO NOT EDIT.

package s3control

import (
	"context"
	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
	"github.com/aws/aws-sdk-go-v2/aws/signer/v4"
	"github.com/aws/aws-sdk-go-v2/service/s3control/types"
	smithy "github.com/awslabs/smithy-go"
	"github.com/awslabs/smithy-go/middleware"
	smithyhttp "github.com/awslabs/smithy-go/transport/http"
)

// This API action puts tags on an Amazon S3 on Outposts bucket. To put tags on an
// S3 bucket, see PutBucketTagging
// (https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketTagging.html) in
// the Amazon Simple Storage Service API. Sets the tags for an Outposts bucket. For
// more information, see Using Amazon S3 on Outposts
// (https://docs.aws.amazon.com/AmazonS3/latest/dev/S3onOutposts.html) in the
// Amazon Simple Storage Service Developer Guide. Use tags to organize your AWS
// bill to reflect your own cost structure. To do this, sign up to get your AWS
// account bill with tag key values included. Then, to see the cost of combined
// resources, organize your billing information according to resources with the
// same tag key values. For example, you can tag several resources with a specific
// application name, and then organize your billing information to see the total
// cost of that application across several services. For more information, see Cost
// Allocation and Tagging
// (https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/cost-alloc-tags.html).
// <note> <p>Within a bucket, if you add a tag that has the same key as an existing
// tag, the new value overwrites the old value. For more information, see <a
// href="https://docs.aws.amazon.com/AmazonS3/latest/dev/CostAllocTagging.html">Using
// Cost Allocation in Amazon S3 Bucket Tags</a>.</p> </note> <p>To use this
// operation, you must have permissions to perform the
// <code>s3outposts:PutBucketTagging</code> action. The Outposts bucket owner has
// this permission by default and can grant this permission to others. For more
// information about permissions, see <a
// href="https://docs.aws.amazon.com/AmazonS3/latest/dev/using-with-s3-actions.html#using-with-s3-actions-related-to-bucket-subresources">
// Permissions Related to Bucket Subresource Operations</a> and <a
// href="https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-access-control.html">Managing
// Access Permissions to Your Amazon S3 Resources</a>.</p> <p>
// <code>PutBucketTagging</code> has the following special errors:</p> <ul> <li>
// <p>Error code: <code>InvalidTagError</code> </p> <ul> <li> <p>Description: The
// tag provided was not a valid tag. This error can occur if the tag did not pass
// input validation. For information about tag restrictions, see <a
// href="https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/allocation-tag-restrictions.html">
// User-Defined Tag Restrictions</a> and <a
// href="https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/aws-tag-restrictions.html">
// AWS-Generated Cost Allocation Tag Restrictions</a>.</p> </li> </ul> </li> <li>
// <p>Error code: <code>MalformedXMLError</code> </p> <ul> <li> <p>Description: The
// XML provided does not match the schema.</p> </li> </ul> </li> <li> <p>Error
// code: <code>OperationAbortedError </code> </p> <ul> <li> <p>Description: A
// conflicting conditional operation is currently in progress against this
// resource. Try again.</p> </li> </ul> </li> <li> <p>Error code:
// <code>InternalError</code> </p> <ul> <li> <p>Description: The service was unable
// to apply the provided tag to the bucket.</p> </li> </ul> </li> </ul> <p>All
// Amazon S3 on Outposts REST API requests for this action require an additional
// parameter of outpost-id to be passed with the request and an S3 on Outposts
// endpoint hostname prefix instead of s3-control. For an example of the request
// syntax for Amazon S3 on Outposts that uses the S3 on Outposts endpoint hostname
// prefix and the outpost-id derived using the access point ARN, see the <a
// href="https://docs.aws.amazon.com/AmazonS3/latest/API/API__control_PutBucketTagging.html#API_control_PutBucketTagging_Examples">
// Example</a> section below.</p> <p>The following actions are related to
// <code>PutBucketTagging</code>:</p> <ul> <li> <p> <a
// href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_control_GetBucketTagging.html">GetBucketTagging</a>
// </p> </li> <li> <p> <a
// href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_control_DeleteBucketTagging.html">DeleteBucketTagging</a>
// </p> </li> </ul>
func (c *Client) PutBucketTagging(ctx context.Context, params *PutBucketTaggingInput, optFns ...func(*Options)) (*PutBucketTaggingOutput, error) {
	stack := middleware.NewStack("PutBucketTagging", smithyhttp.NewStackRequest)
	options := c.options.Copy()
	for _, fn := range optFns {
		fn(&options)
	}
	addawsRestxml_serdeOpPutBucketTaggingMiddlewares(stack)
	awsmiddleware.AddRequestInvocationIDMiddleware(stack)
	smithyhttp.AddContentLengthMiddleware(stack)
	addResolveEndpointMiddleware(stack, options)
	v4.AddComputePayloadSHA256Middleware(stack)
	addRetryMiddlewares(stack, options)
	addHTTPSignerV4Middleware(stack, options)
	awsmiddleware.AddAttemptClockSkewMiddleware(stack)
	addClientUserAgent(stack)
	smithyhttp.AddErrorCloseResponseBodyMiddleware(stack)
	smithyhttp.AddCloseResponseBodyMiddleware(stack)
	addOpPutBucketTaggingValidationMiddleware(stack)
	stack.Initialize.Add(newServiceMetadataMiddleware_opPutBucketTagging(options.Region), middleware.Before)
	addMetadataRetrieverMiddleware(stack)
	addResponseErrorMiddleware(stack)
	v4.AddContentSHA256HeaderMiddleware(stack)

	for _, fn := range options.APIOptions {
		if err := fn(stack); err != nil {
			return nil, err
		}
	}
	handler := middleware.DecorateHandler(smithyhttp.NewClientHandler(options.HTTPClient), stack)
	result, metadata, err := handler.Handle(ctx, params)
	if err != nil {
		return nil, &smithy.OperationError{
			ServiceID:     ServiceID,
			OperationName: "PutBucketTagging",
			Err:           err,
		}
	}
	out := result.(*PutBucketTaggingOutput)
	out.ResultMetadata = metadata
	return out, nil
}

type PutBucketTaggingInput struct {

	// The AWS account ID of the Outposts bucket.
	//
	// This member is required.
	AccountId *string

	// The Amazon Resource Name (ARN) of the bucket. For Amazon S3 on Outposts specify
	// the ARN of the bucket accessed in the format
	// arn:aws:s3-outposts:::outpost//bucket/. For example, to access the bucket
	// reports through outpost my-outpost owned by account 123456789012 in Region
	// us-west-2, use the URL encoding of
	// arn:aws:s3-outposts:us-west-2:123456789012:outpost/my-outpost/bucket/reports.
	// The value must be URL encoded.
	//
	// This member is required.
	Bucket *string

	//
	//
	// This member is required.
	Tagging *types.Tagging
}

type PutBucketTaggingOutput struct {
	// Metadata pertaining to the operation's result.
	ResultMetadata middleware.Metadata
}

func addawsRestxml_serdeOpPutBucketTaggingMiddlewares(stack *middleware.Stack) {
	stack.Serialize.Add(&awsRestxml_serializeOpPutBucketTagging{}, middleware.After)
	stack.Deserialize.Add(&awsRestxml_deserializeOpPutBucketTagging{}, middleware.After)
}

func newServiceMetadataMiddleware_opPutBucketTagging(region string) awsmiddleware.RegisterServiceMetadata {
	return awsmiddleware.RegisterServiceMetadata{
		Region:        region,
		ServiceID:     ServiceID,
		SigningName:   "s3",
		OperationName: "PutBucketTagging",
	}
}
