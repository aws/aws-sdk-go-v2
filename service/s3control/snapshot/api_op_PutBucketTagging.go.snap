PutBucketTagging
	Initialize stack step
		S3Shared:ARNLookup
		S3Control:BackfillInput
		RegisterServiceMetadata
		legacyEndpointContextSetter
		SetLogger
		OperationInputValidation
	Serialize stack step
		stashOperationInput
		ResolveEndpoint
		setArnFieldName
		S3Control:ProcessOutpostIDMiddleware
		S3Control:ProcessARNResourceMiddleware
		OperationSerializer
	Build stack step
		stashOperationInput
		ContentChecksum
		ClientRequestID
		ComputeContentLength
		UserAgent
		RecursionDetection
	Finalize stack step
		ResolveAuthScheme
		GetIdentity
		ResolveEndpointV2
		S3ControlDisableHostPrefix
		disableHTTPS
		EndpointHostPrefix
		ComputePayloadHash
		SigV4ContentSHA256Header
		Retry
		RetryMetricsHeader
		setLegacyContextSigningOptions
		Signing
	Deserialize stack step
		AddRawResponseToMetadata
		ErrorCloseResponseBody
		CloseResponseBody
		ResponseErrorWrapper
		S3MetadataRetriever
		OperationDeserializer
		RecordResponseTiming
		RequestResponseLogger
